src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs

```diff
@@ -4404,78 +4403,134 @@ private void GenerateOneCode()
+                            Ldloc(lenLocal);
+                            Call(s_stringAsSpanIntIntMethod);
+                            Ldc(Operand(0));
+                            Call(s_spanIndexOf);
+                            Stloc(iLocal);
 
-                        if (Code() == RegexCode.Setloop || Code() == RegexCode.Setloopatomic)
-                        {
-                            EmitTimeoutCheck();
-                            EmitMatchCharacterClass(_strings![Operand(0)], IsCaseInsensitive(), charInClassLocal);
-                            BrtrueFar(l1);
+                            Label charFound = DefineLabel();
+
+                            // if (i != -1) goto charFound;
+                            Ldloc(iLocal);
+                            Ldc(-1);
+                            Bne(charFound);
+
+                            // runtextpos += len;
+                            // i = 0;
+                            Ldloc(_runtextposLocal!);
+                            Ldloc(lenLocal);
+                            Add();
+                            Stloc(_runtextposLocal!);
+                            Ldc(0);
+                            Stloc(iLocal);
+                            BrFar(loopEnd);
+
+                            // charFound:
+                            // runtextpos += i;
+                            // i = len - i;
+                            MarkLabel(charFound);
+                            Ldloc(_runtextposLocal!);
+                            Ldloc(iLocal);
+                            Add();
+                            Stloc(_runtextposLocal!);
+                            Ldloc(lenLocal);
+                            Ldloc(iLocal);
+                            Sub();
+                            Stloc(iLocal);
+                            BrFar(loopEnd);
                         }
                         else
                         {
-                            if (IsCaseInsensitive())
+                            // Otherwise, we emit the open-coded loop.
+
+                            Dup();
+                            Stloc(lenLocal);
+                            Ldc(1);
+                            Add();
+                            Stloc(iLocal);
+
+                            Label loopCondition = DefineLabel();
+                            MarkLabel(loopCondition);
+                            Ldloc(iLocal);
+                            Ldc(1);
+                            Sub();
+                            Dup();
+                            Stloc(iLocal);
+                            Ldc(0);
+                            if (Code() == RegexCode.Setloop || Code() == RegexCode.Setloopatomic)
                             {
-                                CallToLower();
+                                BleFar(loopEnd);
+                            }
+                            else
+                            {
+                                Ble(loopEnd);
```

Comment:
Would be nice if we could add an interface to the ILEmitter that would generate the shortest branch possible based on the distance the current instruction is to the target.  It doesn't seem that difficult to do from a cursory glance, but would require some discussion as that's a public interface.

Reply:
https://github.com/dotnet/runtime/issues/32204